File: server/assets/solutions/ohtu2022/part4/python/tdd-ostoskori/src/ostoskori.py
from tuote import Tuote
from ostos import Ostos


class Ostoskori:
    def __init__(self):
        self._ostokset = []

    def _hae_ostos(self, tuote):
        # etsii ensimmäisen kriteerin täyttävän olion listalta, palautta None jos ei löydy
        return next((o for o in self._ostokset if o.tuote.nimi() == tuote.nimi()), None)

    def tavaroita_korissa(self):
        return sum(map(lambda o: o.lukumaara(), self._ostokset))

    def hinta(self):
        return sum(map(lambda o: o.hinta(), self._ostokset))

    def lisaa_tuote(self, lisattava: Tuote):
        ostos = self._hae_ostos(lisattava)
        if ostos:
            ostos.muuta_lukumaaraa(1)
        else:
            self._ostokset.append(Ostos(lisattava))

    def poista_tuote(self, poistettava: Tuote):
        ostos = self._hae_ostos(poistettava)
        if ostos:
            ostos.muuta_lukumaaraa(-1)
        self._ostokset = [o for o in self._ostokset if o.lukumaara() > 0]

    def tyhjenna(self):
        self._ostokset = []

    def ostokset(self):
        return self._ostokset

File: server/assets/solutions/ohtu2021/part4/java/TddOstoskori/src/main/java/ohtu/Ostoskori.java
package ohtu;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class Ostoskori {
    List<Ostos> ostokset;

    public Ostoskori() {
        ostokset = new ArrayList<>();
    }
 
    public int tavaroitaKorissa() {
        return ostokset.stream().mapToInt(o -> o.lukumaara()).sum();
    }
 
    public int hinta() {
        return ostokset.stream().mapToInt(o -> o.hinta()).sum();
    }

    private Ostos haeOstos(Tuote tuote) {
        Optional<Ostos> ostos = this.ostokset.stream()
                .filter(o -> o.tuotteenNimi() == tuote.getNimi())
                .findFirst();

        if (ostos.isPresent()) {
            return ostos.get();
        }
        return null;
    }

    public void lisaaTuote(Tuote lisattava) {
        Ostos ostos = haeOstos(lisattava);
        if (ostos != null) {
            ostos.muutaLukumaaraa(1);
        } else {
            ostokset.add(new Ostos((lisattava)));
        }
    }
 
    public void poista(Tuote poistettava) {
        Ostos ostos = haeOstos(poistettava);
        ostos.muutaLukumaaraa(-1);
        if (ostos.lukumaara() == 0) {
            ostokset = ostokset.stream()
                    .filter(o -> o.tuotteenNimi() != poistettava.getNimi())
                    .collect(Collectors.toList());
        }
    }
 
    public List<Ostos> ostokset() {
        return ostokset;
    }
 
    public void tyhjenna() {
        ostokset = new ArrayList<>();
    }
}
